### PR TITLE
Lukenamop roles patch

### PR DESCRIPTION
**Pull request by lukenamop**
Edit: test/firms.py has been updated and added

Added roles "Associate", "CFO" (Chief Financial Officer), and "COO" (Chief Operating Officer). Updated comment_worker.py, message.py, and models.py to reflect the added roles.

Also updated payroll.py with the added roles. (Might go through and re-optimize payroll.py to better take account for empty roles.)

Ideally, with more roles, the CEO could type something like "!promote lukenamop CFO", but I don't have enough experience in python to figure that out so for now it's just iterative in this order:

Floor Trader:
- no permissions

Associate:
- invite other users to the firm

Executive:
- invite
- promote/demote users to/from associate

CFO (max 1):
- invite
- promote/demote users to/from associate
- change tax rate
- upgrade firm

COO (max 1):
- invite
- promote/demote users to/from executive
- setprivate/setpublic

CEO:
- all permissions

### CODE REVIEW

File: src/comment_worker.py [hunk 1]
```diff
@@ -436,11 +436,26 @@ def firm(self, sess, comment, investor, firm_name=None):
...
         execs = concat_names(
             sess.query(Investor).\
                 filter(Investor.firm == firm.id).\
                 filter(Investor.firm_role == "exec").\
                 all())
+        assocs = concat_names(
+            sess.query(Investor).\
+                filter(Investor.firm == firm.id).\
+                filter(Investor.firm_role == "assoc").\
+                all())
```
Comment (lukenamop) [inline src/comment_worker.py]: Add new roles to the !firm command response

File: src/comment_worker.py [hunk 2]
```diff
@@ -468,14 +489,20 @@ def firm(self, sess, comment, investor, firm_name=None):
...
         else:
             return comment.reply_wrap(
                 message.modify_firm_other(
                     firm,
                     ceo,
+                    coo,
+                    cfo,
                     execs,
+                    assocs,
                     traders))
```
Comment (lukenamop) [inline src/comment_worker.py:510]: Lines 439 through 506: Added new roles to the "!firm" command response

File: src/comment_worker.py [hunk 3]
```diff
@@ -532,13 +559,20 @@ def leavefirm(self, sess, comment, investor):
...
         firm.size -= 1
-        if investor.firm_role == 'exec':
+        if investor.firm_role == 'coo':
+            firm.coo -= 1
+        elif investor.firm_role == 'cfo':
+            firm.cfo -= 1
+        elif investor.firm_role == 'exec':
             firm.execs -= 1
+        elif investor.firm_role == 'assoc':
+            firm.assocs -= 1
```
Comment (lukenamop) [inline src/comment_worker.py:579]: Lines 567 through 575: Added new roles to the "!leavefirm" command

File: src/comment_worker.py [hunk 4]
```diff
@@ -603,33 +688,56 @@ def fire(self, sess, comment, investor, to_fire):
...
         user = sess.query(Investor).\
             filter(func.lower(Investor.name) == func.lower(to_fire)).\
             first()
         if (user == None) or (user.name == investor.name) or (user.firm != investor.firm):
             return comment.reply_wrap(message.fire_failure_org)
 
-        if (investor.firm_role != "ceo") and (user.firm_role != ""):
-            return comment.reply_wrap(message.not_ceo_org)
+        #if (investor.firm_role != "ceo") and (user.firm_role != ""):
+        #    return comment.reply_wrap(message.not_ceo_org)
```
Comment (lukenamop) [inline src/comment_worker.py:702]: Lines 697 and 698 can be deleted

File: src/comment_worker.py [hunk 5]
```diff
@@ -568,28 +599,82 @@ def promote(self, sess, comment, investor, to_promote):
...
             flair_role_investor = "CEO"
-        else:
+        elif investor.firm_role == "coo":
+            flair_role_investor = "COO"
+        elif investor.firm_role == "cfo":
+            flair_role_investor = "CFO"
+        elif investor.firm_role == "exec":
             flair_role_investor = "Executive"
+        elif investor.firm_role == "assoc":
+            flair_role_investor = "Associate"
```
Comment (lukenamop) [inline src/comment_worker.py:681]: Lines 590 through 677: Added new roles to "!promote" command as well as adjusted role permissions

File: src/comment_worker.py [hunk 6]
```diff
@@ -603,33 +688,56 @@ def fire(self, sess, comment, investor, to_fire):
...
+
+            firm.cfo -= 1
+
+        elif user.firm_role == "coo":
+            if (investor.firm_role != "ceo"):
+                return comment.reply_wrap(message.not_ceo_org)
+
+            firm.coo -= 1
+
+        firm.size -= 1
```
Comment (lukenamop) [inline src/comment_worker.py:736]: Lines 690 through 732: Added new roles to "!fire" command as well as adjusted role permissions

File: src/comment_worker.py [hunk 7]
```diff
@@ -672,7 +780,7 @@ def invite(self, sess, comment, investor, invitee_name):
             return comment.reply_wrap(message.no_firm_failure_org)
 
         if investor.firm_role == "":
-            return comment.reply_wrap(message.not_ceo_or_exec_org)
+            return comment.reply_wrap(message.not_assoc_org)
```
Comment (lukenamop) [inline src/comment_worker.py:787]: Changed "!invite" permissions

File: src/comment_worker.py [hunk 8]
```diff
@@ -698,8 +806,8 @@ def setprivate(self, sess, comment, investor):
         if investor.firm == 0:
             return comment.reply_wrap(message.no_firm_failure_org)
 
-        if investor.firm_role != "ceo":
-            return comment.reply_wrap(message.not_ceo_org)
+        if (investor.firm_role != "ceo") or (investor.firm_role != "coo"):
+            return comment.reply_wrap(message.not_ceo_or_coo_org)
```
Comment (lukenamop) [inline src/comment_worker.py:814]: Changed "!setprivate" permissions

File: src/comment_worker.py [hunk 9]
```diff
@@ -714,25 +822,25 @@ def setpublic(self, sess, comment, investor):
...
```
Comment (lukenamop) [inline src/comment_worker.py:830]: Changed "!setpublic" permissions

File: src/comment_worker.py [hunk 10]
```diff
@@ -714,25 +822,25 @@ def setpublic(self, sess, comment, investor):
...
 
         firm = sess.query(Firm).\
             filter(Firm.id == investor.firm).\
             first()
 
         firm.private = False
 
-        return comment.reply_wrap(message.setprivate_org)
+        return comment.reply_wrap(message.setpublic_org)
```
Comment (lukenamop) [inline src/comment_worker.py:838]: Fixed the "!setpublic" error message (was incorrectly set to setprivate_org)

File: src/comment_worker.py [hunk 11]
```diff
@@ -714,25 +822,25 @@ def setpublic(self, sess, comment, investor):
...
     @req_user
     def tax(self, sess, comment, investor, tax_temp):
 
         if investor.firm == 0:
             return comment.reply_wrap(message.no_firm_failure_org)
 
-        if investor.firm_role != "ceo":
-            return comment.reply_wrap(message.not_ceo_org)
+        if (investor.firm_role != "ceo") or (investor.firm_role != "cfo"):
+            return comment.reply_wrap(message.not_ceo_or_cfo_org)
```
Comment (lukenamop) [inline src/comment_worker.py:847]: Changed "!tax" permissions

File: src/comment_worker.py [hunk 12]
```diff
@@ -759,8 +867,8 @@ def upgrade(self, sess, comment, investor):
         if investor.firm == 0:
             return comment.reply_wrap(message.nofirm_failure_org)
 
-        if investor.firm_role != "ceo":
-            return comment.reply_wrap(message.not_ceo_org)
+        if (investor.firm_role != "ceo") or (investor.firm_role != "cfo"):
+            return comment.reply_wrap(message.not_ceo_or_cfo_org)
```
Comment (lukenamop) [inline src/comment_worker.py:875]: Changed "!upgrade" permissions

File: src/comment_worker.py [hunk 13]
```diff
@@ -779,8 +887,9 @@ def upgrade(self, sess, comment, investor):
 
         max_members = max_members_for_rank(firm.rank)
         max_execs = max_execs_for_rank(firm.rank)
+        max_assocs = max_assocs_for_rank(firm.rank)
 
-        return comment.reply_wrap(message.modify_upgrade(firm, max_members, max_execs))
+        return comment.reply_wrap(message.modify_upgrade(firm, max_members, max_execs, max_assocs))
```
Comment (lukenamop) [inline src/comment_worker.py:896]: Added new roles to "!updrade"

File: src/comment_worker.py [hunk 14]
```diff
@@ -793,9 +902,16 @@ def max_members_for_rank(rank):
     # etc.
     return 2 ** (rank + 3)
 
+def max_assocs_for_rank(rank):
+    # level 1 = 2
+    # level 2 = 6
+    # level 3 = 14
+    # etc.
+    return ((2 ** (rank + 3)) / 2) - 2
+
```
Comment (lukenamop) [inline src/comment_worker.py:915]: Added function to return max number of associates for a firm of given rank

File: src/message.py [hunk 15]
```diff
@@ -386,13 +386,34 @@ def modify_pay_to_post(balance):
...
+
+*CFO:*
+%CFO%
+
 *Executives:*
 %EXECS%
 
+*Associates:*
+%ASSOCS%
+
```
Comment (lukenamop) [inline src/message.py]: Added new roles to "!firm" command response

File: src/message.py [hunk 16]
```diff
@@ -420,36 +450,33 @@ def modify_pay_to_post(balance):
...
+rank_strs = {
+    "ceo": "CEO",
+    "coo": "COO",
+    "cfo": "CFO",
+    "exec": "Executive",
+    "assoc": "Associate",
+    "": "Floor Trader"
+}
+
```
Comment (lukenamop) [inline src/message.py:479]: Lines 389 through 479: Added new roles to "!firm" command response

File: src/message.py [hunk 17]
```diff
@@ -486,7 +513,7 @@ def modify_createfirm_exists_failure(firm_name):
 leavefirm_ceo_failure_org = """
 You are currently the CEO of your firm, so you are not allowed to leave.
 
-If you really want to leave, you will need to first demote yourself by promoting an executive member to CEO with the **!promote <username>** command.
+If you really want to leave, you will need to first demote yourself by promoting a COO or CFO member to CEO with the **!promote <username>** command.
```
Comment (lukenamop) [inline src/message.py:516]: Adjusted response to a CEO calling "!leavefirm"

File: src/message.py [hunk 18]
```diff
@@ -497,45 +524,74 @@ def modify_createfirm_exists_failure(firm_name):
...
+Could not promote this employee since the firm already has a CFO.
+"""
+
+promote_execs_full_org = """
+Could not promote this employee since the firm is at its maximum executive limit.
+**Number of executives:** %EXECS%
 **Firm level:** %LEVEL%
 
-The CEO of the firm can raise this limit by upgrading with `!upgrade`.
+The CEO or CFO of the firm can raise this limit by upgrading with `!upgrade`.
```
Comment (lukenamop) [inline src/message.py:556]: Lines 527 through 556: Adjusted and created error messages to reflect new role permissions

File: src/message.py [hunk 19]
```diff
@@ -497,45 +524,74 @@ def modify_createfirm_exists_failure(firm_name):
...
 """
 
 def modify_fire(user):
     return fire_org.\
         replace("%NAME%", user.name)
 
-fire_org = """
-Successfully fired **/u/%NAME%** from the firm.
-"""
-
```
Comment (lukenamop) [inline src/message.py:538]: Moved "fire_org" above "modify_fire" and "promote_org" above "modify_promote" for sake of clarity

File: src/message.py [hunk 20]
```diff
@@ -497,45 +524,74 @@ def modify_createfirm_exists_failure(firm_name):
...
+**Number of associates:** %ASSOCS%
+**Firm level:** %LEVEL%
+
+The CEO or CFO of the firm can raise this limit by upgrading with `!upgrade`.
+"""
+
+def modify_promote_assocs_full(firm):
+    return promote_assocs_full_org.\
+        replace("%ASSOCS%", str(firm.assocs)).\
+        replace("%LEVEL%", str(firm.rank + 1))
```
Comment (lukenamop) [inline src/message.py:575]: Lines 559 through 575: Changed "!promote" error messages to reflect difference between associate limit and executive limit

File: src/message.py [hunk 21]
```diff
@@ -663,14 +719,15 @@ def modify_upgrade_insufficient_funds_org(firm, cost):
 upgrade_org = """
 You have succesfully upgraded the firm to **level %LEVEL%**!
 
-The firm may now have up to **%MAX_MEMBERS% employees**, including up to **%MAX_EXECS% executives**.
+The firm may now have up to **%MAX_MEMBERS% employees**, including up to **%MAX_EXECS% executives** and **%MAX_ASSOCS% associates**.
 """
 
-def modify_upgrade(firm, max_members, max_execs):
+def modify_upgrade(firm, max_members, max_execs, max_assocs):
```
Comment (lukenamop) [inline src/message.py:725]: Lines 606 through 725: Changed error messages to reflect new role permissions

File: src/models.py [hunk 22]
```diff
@@ -71,7 +71,10 @@ class Firm(Base):
     name = Column(String(32), nullable=False, unique=True)
     balance = Column(BigInteger, default=1000)
     size = Column(Integer, default=0)
+    coo = Column(Integer, default=0)
+    cfo = Column(Integer, default=0)
     execs = Column(Integer, default=0)
+    assocs = Column(Integer, default=0)
```
Comment (lukenamop) [inline src/models.py:77]: Added new roles to Firm class

File: src/payroll.py [hunk 23]
```diff
@@ -47,32 +47,59 @@ def main():
             # 10% of firm coins are burned as a tax
             firm.balance -= int(0.1 * firm.balance)
 
-            payout_amount = int(0.4 * firm.balance)
+            # 50% of remaining firm coins are paid out
+            payout_amount = int(0.5 * firm.balance)
```
Comment (lukenamop) [inline src/payroll.py:51]: Slightly adjusted total payout_amount

File: src/payroll.py [hunk 24]
```diff
@@ -47,32 +47,59 @@ def main():
...
             if payout_amount == 0:
                 # handle broke firms
                 firm.last_payout = now
                 continue
 
+            # 30% paid to board members (CEO, COO, CFO) (30% of total payroll)
+            board_total = payout_amount * 0.3
+            board_members = 1 + firm.coo + firm.cfo
+            board_amount = int(board_total / board_members)
```
Comment (lukenamop) [inline src/payroll.py:60]: Added "board_payout": money paid to CEO, COO, and CFO

File: src/payroll.py [hunk 25]
```diff
@@ -47,32 +47,59 @@ def main():
...
+                remaining_amount -= assoc_total
+
+            # 100% of remaining paid to associates (21% of total payroll)
+            trader_total = remaining_amount
             trader_amount = int(trader_total / max(firm.size - firm.execs, 1))
 
-            logging.info(" -- firm '%s': paying out %s each to %s traders, and %s each to %s execs",\
-                firm.name, trader_amount, firm.size - firm.execs, exec_amount, firm.execs)
+            logging.info(" -- firm '%s': paying out %s each to %s trader(s), %s each to %s associate(s), %s each to %s executive(s), and %s each to %s board member(s)",\
+                firm.name, trader_amount, firm.size - firm.execs, assoc_amount, firm.assocs, exec_amount, firm.execs, board_amount, board_members)
```
Comment (lukenamop) [inline src/payroll.py:85]: Lines 62 through 85: Adjusted payroll to reflect new roles (view the comments on lines 57, 64, 72, and 80 to see new payroll amounts)

**Comment by mappum** [comment]
@lukenamop Will you be updating the unit tests? (`/test/firms.py`). You can run them with `./test.sh`.

**Comment by lukenamop** [comment]
@mappum I'm new to github & new to working with this project so I didn't know that was a thing I should do! I'll get started now. Thanks :)

**Comment by mappum** [comment]
> I'm new to github & new to working with this project so I didn't know that was a thing I should do! I'll get started now. Thanks :)

No worries, lmk if you have questions writing the tests. It should end up being much less work than manually testing each code path by hand.

**Comment by lukenamop** [comment]
@mappum I do have one question so far. In firms.py, sometimes the test user is referred to as "foo" and sometimes as "testuser". Does it matter which one I use? Like, to check the investor, should I be using "Investor.name == 'testuser'" or "Investor.name == 'foo'"?

**Comment by lukenamop** [comment]
@mappum Almost halfway done... I think xD

Are there any other files in the test folder I should check to make sure they will still work? If you aren't sure I'll browse through them all and change anything I notice.

**Comment by mappum** [comment]
@lukenamop Doesn't seem like you touched much outside of firm-specific logic, so if they are all passing I'd assume things are OK.

**Comment by mappum** [comment]
Oh and protip: you can open `htmlcov/index.html` after running the tests to see which lines are covered. (It doesn't have to be 100%, but it's nice for knowing if you tested a specific condition or not).

**Comment by lukenamop** [comment]
> Oh and protip: you can open `htmlcov/index.html` after running the tests to see which lines are covered. (It doesn't have to be 100%, but it's nice for knowing if you tested a specific condition or not).

Thanks! I'll try it out when I get through writing these tests.

**Comment by lukenamop** [comment]
@mappum Okay, I think I made all the tests I need. How can I run them?

**Comment by mappum** [comment]
@lukenamop `./test.sh`

**Comment by lukenamop** [comment]
@mappum I've been trying for a while and can't get it to cooperate. 
[memeinvestor terminal log.txt](https://github.com/MemeInvestor/memeinvestor_bot/files/2933683/memeinvestor.terminal.log.txt)

**Comment by lukenamop** [comment]
@mappum The test / firms.py file should be all set to go though if you could test it :)

File: src/comment_worker.py [hunk 26]
```diff
@@ -450,13 +465,23 @@ def firm(self, sess, comment, investor, firm_name=None):
...
+            elif investor.firm_role == "coo":
+                flair_role = "COO"
+            elif investor.firm_role == "cfo":
+                flair_role = "CFO"
+            elif investor.firm_role == "exec":
+                flair_role = "Executive"
+            elif investor.firm_role == "assoc":
+                flair_role = "Associate"
+            else:
+                flair_role = "Floor Trader"
```
Comment (lukenamop) [inline src/comment_worker.py:484]: Lines 468 through 484: Fixed incorrect firm flair assignment when calling the "!firm (other)" command

**Comment by lukenamop** [comment]
@mappum @thecsw Have you guys had a chance to test this pull request yet?

**Comment by mappum** [comment]
@lukenamop we'd like to adopt your changes, can you message me on Discord? My username is `mappum#0804`

**Pull request closed without merging**